### PR TITLE
Backward weight atomic add(fp32)

### PR DESCRIPTION
**Pull request by ltqin**
This is a backward weight using atomic add ,  and is also only for fp32

Note:
1.the way of calculating kblocks is different from miopen. because "miopen-affix-params " is after "miopen-lowering". it may effect the performance

### CODE REVIEW

File: mlir/lib/Dialect/MIOpen/Transforms/AffineTransforms.cpp [hunk 1]
```diff
@@ -108,7 +108,8 @@ AffineMap AffineTransforms::buildIndexAffineMap(miopen::TransformOp op) {
           auto srcDim = srcDimAttr.getValue()[j].cast<IntegerAttr>().getInt();
           affExprsMap.insert({srcDim, expr});
         }
-      } else if (transformAttr.getValue() == "UnMerge") {
+      } else if ((transformAttr.getValue() == "UnMerge2") ||
```
Comment (whchung) [inline mlir/lib/Dialect/MIOpen/Transforms/AffineTransforms.cpp]: @ltqin the affine map for `UnMerge` and `UnMerge2` seems to be identical?

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 2]
```diff
@@ -4465,163 +4506,1149 @@ struct Conv2DRewritePattern : public OpRewritePattern<T> {
...
+    // compute padding hi/wi.
+    auto hiPadded = hi + leftPadH + rightPadH;
+    auto wiPadded = wi + leftPadW + rightPadW;
+    // calculate gemmKBlocks
+    int64_t gemmN = c * y * x;
+    int64_t gemmM = k;
+    int64_t gemmK = n * ho * wo;
+    int64_t standardBockNum = ((gemmN + 127) / 128) * ((gemmM + 127) / 128);
+    // static const int64_t MaxSubBlockNum = 2048 / standardBockNum;
+    int64_t gemmKBlocks = 1;
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h]: logic for `gemmKBlocks` seem to be duplicated somehow?

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 3]
```diff
@@ -702,6 +714,30 @@ computeIndexDiffMap(OpBuilder &b, Location loc,
...
+      int64_t lowerDim = q[0].template cast<IntegerAttr>().getInt();
+      lowerIndicesDiffMap[lowerDim] = lowerDiff;
+      lowerIndicesUpdatedMap[lowerDim] = b.create<AddIOp>(
+          loc,
+          b.create<IndexCastOp>(loc, lowerIndicesOriginal[lowerDim],
+                                b.getIntegerType(32)),
+          lowerDiff);
+    } else if (transformation.getValue() == "UnMerge2") {
+      auto e = g.get("parameters").template cast<ArrayAttr>();
+      assert(e.size() == p.size());
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:920]: it seems the only difference between `UnMerge` and `UnMerge2` is that `UnMerge2` can have more than 2 upper dimensions?

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 4]
```diff
@@ -4465,163 +4506,1149 @@ struct Conv2DRewritePattern : public OpRewritePattern<T> {
...
+            b.getNamedAttr(
+                "dimension_lengths",
+                b.getArrayAttr({b.getI32IntegerAttr(gemmKBlocks),
+                                b.getI32IntegerAttr(n / gemmKBlocks)})),
+            b.getNamedAttr("transformation", b.getStringAttr("UnMerge2")),
+            b.getNamedAttr("parameters",
+                           b.getArrayAttr({
+                               b.getI32IntegerAttr(gemmKBlocks),
+                               b.getI32IntegerAttr(n / gemmKBlocks),
+                           })),
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h]: from the look how `UnMerge2` is used, there are only two parameters, so wondering why `UnMerge` can not be used?

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 5]
```diff
@@ -327,7 +329,17 @@ inline void emitStoreLogic(OpBuilder &b, Location loc, MemRefType destType,
...
+    if (memoryOp == InMemoryDataOperation::Set)
+      b.create<StoreOp>(loc, value, dest, destLowerIndices);
+    else {
+      SmallVector<Value, 8> destLowerStoreIndices;
+      for (unsigned i = 0; i < destLowerIndices.size(); ++i) {
+        auto dstIndex = b.create<IndexCastOp>(loc, destLowerIndices[i],
+                                              b.getIntegerType(32));
+        destLowerStoreIndices.push_back(dstIndex);
+      }
+      b.create<gpu::AtomicFAddOp>(loc, value, dest, destLowerStoreIndices);
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:535]: `atomicFAdd` is only used in non-OOB cases in the current implementation, would it be used in OOB cases?

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 6]
```diff
@@ -1152,8 +1188,17 @@ struct Conv2DRewritePattern : public OpRewritePattern<T> {
...
+    auto xdlopsV2Attr = op->template getAttrOfType<BoolAttr>("xdlopsV2");
+    if (xdlopsV2Attr && xdlopsV2Attr.getValue() == true)
+      isXdlops = true;
+    auto dataType =
+        op.input().getType().template cast<MemRefType>().getElementType();
     if (miopen::ConvOpType::Conv2DBwdDataOpType == convOpType) {
       return backwardData(op, b);
+    } else if (miopen::ConvOpType::Conv2DBwdWeightOpType == convOpType &&
+               isXdlops && dataType == b.getF32Type()) {
+      return backwardWeightAtomicAdd(op, b);
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h]: comments are needed to indicate currently for backward weight we only do it in XDLOPS + f32 path.

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 7]
```diff
@@ -4465,163 +4506,1149 @@ struct Conv2DRewritePattern : public OpRewritePattern<T> {
...
-    a_block_space = math::integer_least_multiple(KPerBlock * AlignedMPerBlock,
-                                                 max_lds_align);
+    // compute padding hi/wi.
+    auto hiPadded = hi + leftPadH + rightPadH;
+    auto wiPadded = wi + leftPadW + rightPadW;
+    // calculate gemmKBlocks
+    int64_t gemmN = c * y * x;
+    int64_t gemmM = k;
+    int64_t gemmK = n * ho * wo;
+    int64_t standardBockNum = ((gemmN + 127) / 128) * ((gemmM + 127) / 128);
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h]: `Bock` -> `Block`?

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 8]
```diff
@@ -7563,6 +8594,13 @@ struct ThreadwiseCopyV2RewritePattern
 
     auto sourceType = op.source().getType().cast<VectorType>();
     auto destType = op.dest().getType().cast<MemRefType>();
+    InMemoryDataOperation dataOpration = InMemoryDataOperation::Set;
+    auto dataOpAttr = op->getAttr("data_operation");
+    if (dataOpAttr) {
+      int64_t dataOp = dataOpAttr.template cast<IntegerAttr>().getInt();
+      if (dataOp == 1)
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:8601]: use the enum?

Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:8363]: `data_operation = 1` is the attribute specifically used in backward weight + XDLOPS + f32 path, right?

File: mlir/lib/Dialect/MIOpen/Transforms/AffineTransforms.cpp (hunk 1 above)
Comment (ltqin) [reply mlir/lib/Dialect/MIOpen/Transforms/AffineTransforms.cpp]: computing diff is different, the diff of UnMerge is Embed, not UnMerge

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 3 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:920]: yes, UnMerge2 can have more than 2 upper dimensions

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 5 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:535]: for backward weight, OOB is not required

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 8 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:8363]: data_operation = 1 mean using atomic add. so upper level should know what time(condition) to using it

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 7 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h]: yes, I will remove it

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 3 above)
Comment (whchung) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:920]: Can we merge `UnMerge` and `UnMerge2` in such case? The only thing we need to do is to relax the assert in `UnMerge`? Actually I think the rule of upper dimensions be 2 is only required in `Embed`?

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 2 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h]: done

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 6 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h]: done

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 3 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:920]: the algorithm of UnMerge2 is different current UnMerge

**Review by whchung — CHANGES_REQUESTED** [review]
@ltqin per discussion, let's do this:

- rename `UnMerge` to `Embed`
- rename `UnMerge2` to `UnMerge`

please also supply comments where index diffs are computed for the new `UnMerge`.

**Comment by ltqin** [comment]
@whchung comment of UnMerge is right, so I not change it

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 9]
```diff
@@ -1335,20 +1370,26 @@ inline void computeTopAndBottomIndicesWithAffineMap(
 //===----------------------------------------------------------------------===//
 
 // The ArgumentFields keep track of differences between conv operations
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h]: this comment wrt `ArgumentFields` should be moved or eliminated along with `ArgumentFields`.

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 8 above)
Comment (whchung) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:8363]: a comment would be nice.

File: mlir/include/mlir/Dialect/MIOpen/utility/common.hpp [hunk 10]
```diff
@@ -0,0 +1,8 @@
+#ifndef COMMON_H
+#define COMMON_H
+struct ArgumentFields {
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/utility/common.hpp]: please move the comment of `ArgumentFields` here.

File: mlir/include/mlir/Dialect/MIOpen/utility/common.hpp [hunk 11]
```diff
@@ -0,0 +1,8 @@
+#ifndef COMMON_H
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/utility/common.hpp]: please rename the file to `common.h`.

same with `math.hpp`. please help rename it to `math.h` to cope with LLVM / MLIR naming convention.

**Review by whchung — CHANGES_REQUESTED** [review]
some minor changes and I think the PR is good to go.

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 9 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h]: done

File: mlir/include/mlir/Dialect/MIOpen/utility/common.hpp (hunk 10 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/utility/common.hpp]: done

File: mlir/include/mlir/Dialect/MIOpen/utility/common.hpp (hunk 11 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/utility/common.hpp]: done

**Review by whchung — DISMISSED** [review]

**Comment by whchung** [comment]
@ltqin CI test show this PR would cause memory access fault in one of the resnet50 config:

```
mlir-miopen-driver -fil_layout=kcyx -in_layout=nchw -out_layout=nkhw -batchsize=256 -in_channels=128 -in_h=28 -in_w=28 -out_channels=128 -fil_h=3 -fil_w=3 --dilation_h=1 --dilation_w=1 --conv_stride_h=1 --conv_stride_w=1 --padding_h=1 --padding_w=1 -t f16 --operation conv2d_bwd_weight -pv -x2 -c | mlir-rocm-runner --shared-libs=/home/jenkins/workspace/mlir_PR-271/build/lib/librocm-runtime-wrappers.so,/home/jenkins/workspace/mlir_PR-271/build/lib/libmlir_runner_utils.so --entry-point-result=void | /home/jenkins/workspace/mlir_PR-271/build/bin/FileCheck /home/jenkins/workspace/mlir_PR-271/mlir/test/mlir-miopen-driver/e2e_for_pr/Resnet50_validation_nchw.mlir --check-prefix=CHECK_NCHW_3
```

**Comment by ltqin** [comment]
@whchung I forget fp16 not using atomic add

**Review by whchung — APPROVED** [review]

**Pull request merged**